metadata/src/main/java/org/apache/kafka/metadata/util/BatchFileReader.java

```diff
@@ -0,0 +1,183 @@
+        for (Iterator<Record> iter = input.iterator(); iter.hasNext(); ) {
+            Record record = iter.next();
+            try {
+                short typeId = ControlRecordType.parseTypeId(record.key());
+                ControlRecordType type = ControlRecordType.fromTypeId(typeId);
+                switch (type) {
+                    case LEADER_CHANGE: {
+                        LeaderChangeMessage message = new LeaderChangeMessage();
+                        message.read(new ByteBufferAccessor(record.value()), (short) 0);
+                        messages.add(new ApiMessageAndVersion(message, (short) 0));
+                        break;
+                    }
+                    case SNAPSHOT_HEADER: {
+                        SnapshotHeaderRecord message = new SnapshotHeaderRecord();
+                        message.read(new ByteBufferAccessor(record.value()), (short) 0);
+                        messages.add(new ApiMessageAndVersion(message, (short) 0));
+                        break;
+                    }
+                    case SNAPSHOT_FOOTER: {
+                        SnapshotFooterRecord message = new SnapshotFooterRecord();
+                        message.read(new ByteBufferAccessor(record.value()), (short) 0);
+                        messages.add(new ApiMessageAndVersion(message, (short) 0));
+                        break;
+                    }
+                    default:
+                        throw new RuntimeException("Unsupported control record type " + type + " at offset " +
+                                record.offset());
+                }
+            } catch (Throwable e) {
+                throw new RuntimeException("Unable to read control record at offset " + record.offset(), e);
+            }
+        }
+        return new BatchAndType(Batch.data(
+            input.baseOffset(),
+            input.partitionLeaderEpoch(),
+            input.maxTimestamp(),
+            input.sizeInBytes(),
+            messages), true);
+    }
+
+    private BatchAndType nextMetadataBatch(FileChannelRecordBatch input) {
+        List<ApiMessageAndVersion> messages = new ArrayList<>();
+        for (Record record : input) {
+            ByteBufferAccessor accessor = new ByteBufferAccessor(record.value());
+            try {
+                ApiMessageAndVersion messageAndVersion = serde.read(accessor, record.valueSize());
+                messages.add(messageAndVersion);
+            } catch (Throwable e) {
+                log.error("unable to read metadata record at offset {}", record.offset(), e);
+            }
+        }
+        return new BatchAndType(Batch.data(
+            input.baseOffset(),
+            input.partitionLeaderEpoch(),
+            input.maxTimestamp(),
+            input.sizeInBytes(),
+            messages), false);
+    }
+
+    @Override
+    public void close() {
+        try {
+            fileRecords.closeHandlers();
```

Comment:
`FileRecords#close` tries to do a flush and trim, which fails because the file descriptor was opened read-only.

`FileRecords#close` probably should be fixed to remove the assumption that all files are writable, but I didn't want to expand the scope of this change even more.